docs/architecture/adr-030-authz-module.md

```diff
@@ -0,0 +1,222 @@
+implementation is based on work done by the [Gaian's team at Hackatom Berlin 2019](https://github.com/cosmos-gaians/cosmos-sdk/tree/hackatom/x/delegation).
+
+## Decision
+
+We will create a module named `authz` which provides support for
+granting arbitrary capabilities from one account (the granter) to another account (the grantee). Authorizations
+must be granted for a particular `Msg` service methods one by one using an implementation
+of `Authorization`.
+
+### Types
+
+Authorizations determine exactly what capabilities are granted. They are extensible
+and can be defined for any `Msg` service method even outside of the module where
+the `Msg` method is defined. `Authorization`s use the new `ServiceMsg` type from
+ADR 031.
+
+#### Authorization
+
+```go
+type Authorization interface {
+	// MethodName returns the fully-qualified Msg service method name as described in ADR 031.
+	MethodName() string
+
+	// Accept determines whether this grant permits the provided sdk.ServiceMsg to be performed, and if
+	// so provides an upgraded authorization instance. If delete returns true,the Authorization has
+	// been exhausted and can be deleted from state.
+	Accept(msg sdk.ServiceMsg, block abci.Header) (allow bool, updated Authorization, delete bool)
+}
+```
+
+For example a `SendAuthorization` like this is defined for `MsgSend` that takes
+a `SpendLimit` and updates it down to zero:
+
+```go
+type SendAuthorization struct {
+	// SpendLimit specifies the maximum amount of tokens that can be spent
+	// by this authorization and will be updated as tokens are spent. If it is
+	// empty, there is no spend limit and any amount of coins can be spent.
+	SpendLimit sdk.Coins
+}
+
+func (cap SendAuthorization) MethodName() string {
+	return "/cosmos.bank.v1beta1.Msg/Send"
+}
+
+func (cap SendAuthorization) Accept(msg sdk.ServiceMsg, block abci.Header) (allow bool, updated Authorization, delete bool) {
+	switch req := msg.Request.(type) {
+	case bank.MsgSend:
+		left, invalid := cap.SpendLimit.SafeSub(req.Amount)
+		if invalid {
+			return false, nil, false
+		}
+		if left.IsZero() {
+			return true, nil, true
+		}
+		return true, SendAuthorization{SpendLimit: left}, false
+	}
+	return false, nil, false
+}
+```
+
+A different type of capability for `MsgSend` could be implemented
+using the `Authorization` interface with no need to change the underlying
+`bank` module.
+
+### `Msg` Service
+
+```proto
+service Msg {
+  // GrantAuthorization grants the provided authorization to the grantee on the granter's
+  // account with the provided expiration time.
+  rpc GrantAuthorization(MsgGrantAuthorization) returns (MsgGrantAuthorizationResponse);
+
+  // ExecAuthorized attempts to execute the provided messages using
+  // authorizations granted to the grantee. Each message should have only
+  // one signer corresponding to the granter of the authorization.
+  rpc ExecAuthorized(MsgExecAuthorized) returns (MsgExecAuthorizedResponse)
```

Comment:
Could you specify how this method will call other services? Let's use the `SendAuthorization` example. Will `x/authz` have clients to all other modules? How it will inform other module that a message is authorized?

I thought that `DispatchActions` will do that. But it seams that the `DispatchActions` is used for other modules to call `x/authz`.

Reply:
I think we addressed this when we chatted right?

Reply:
yes, I can add more details in other PR.